code/modules/atmospherics/gasmixtures/gas_mixture.dm

```diff
@@ -183,9 +189,17 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 
 	var/list/cached_gases = gases //accessing datum vars is slower than proc vars
 	var/list/giver_gases = giver.gases
+	var/list/gases_both = cached_gases & giver_gases
+	var/list/gases_new = giver_gases ^ gases_both
+
+	var/list/tmp_gas
+	//add new gases
+	for(var/new_id in gases_new)
+		tmp_gas = gases_new[new_id]
+		cached_gases[new_id] = tmp_gas.Copy() //from ADD_GAS
+
 	//gas transfer
-	for(var/giver_id in giver_gases)
-		ASSERT_GAS(giver_id, src)
+	for(var/giver_id in gases_both)
 		cached_gases[giver_id][MOLES] += giver_gases[giver_id][MOLES]

```

Comment:
What think about this?

Reply:
As I understand it merge() is called under the assumption that giver can be discarded afterwords.

The code you're replacing here is faster then or as fast as what you've replaced it with.
For one we have to spend time creating 3 new lists.
We then have to add all the gases together, not an expensive task. 

The painful part is the ^. As far as I can tell this means we have to remove all the gases that are in both and giver. So we are left with the gases in both that were not from giver, namely the gases from cashed.

This means that the for loop is just adding the gases that are in cashed back into cashed.

This is still the same or worse speed then ASSERT_GAS, which looks like this:
`if (!gas_mixture.gases[gas_id]) { ADD_GAS(gas_id, gas_mixture.gases) };`
That's 1 if check and an array access.
ADD_GAS looks like this:
`var/list/tmp_gaslist = GLOB.gaslist_cache[gas_id]; out_list[gas_id] = tmp_gaslist.Copy();`
Which is an array access and a copy().

Don't do this here.

Reply:
first i think replace part of ASSERT_GAS to ADD_GAS 
and next replace part with copy of global  and add moles to copy giver gas

>We then have to add all the gases together, not an expensive task.
wrong sentence, 
`gases_both = cached_gases & giver_gases` not summ. the result is a list that contains only items that were in both lists. so this is gases that need only mole change. no need assert.

next i remove this gases from list of new gases so i get list of gases that need ADD_GAS.

about creating of lists, what if put `cached_gases & giver_gases` directly in for, insteat of new var?

Reply:
Ah I understand, forgive me.

Putting it in the for loop would help, I'm still not quite sure why you're using this over ASSERT_GAS.
Is there something I'm not getting?

Reply:
assert_gas use check, i think bypas checks can lower load. 
I dont wery sure in this change but if it remove some load why not implement it?

Reply:
I don't think your current method is cheaper then the assert_gas define. It goes about making the check for the gas in a different way but it does around the same thing. I can't say for sure because I don't know precisely how byond does ^ and & on lists.

It would be a good idea to profile it but you should probably do that on a clean master so it doesn't get effected by any of the other changes you've made here.

Reply:
I think first better create clear byond test build to check this operations.

Reply:
From /datum/gas_mixture/proc/share(datum/gas_mixture/sharer, atmos_adjacent_turfs = 4)
https://github.com/tgstation/tgstation/blob/26d1edbe81f66ff52e8c4fd32bc55885dd4bc86a/code/modules/atmospherics/gasmixtures/gas_mixture.dm#L278-L281